File: UQ-Packages/wavegen.py
# coding: utf-8
# WaveMaker Code
# by kense
# ############
import numpy as np
import matplotlib.pyplot as plt

np.random.seed(seed=None) # random seed

def jonswap(f,Hs,Tp): # JONSWAP function

	fp=1/float(Tp)

	TH = Tp/float(np.sqrt(Hs))

	if TH<=3.6:

			gamma=5

	elif (3.6<TH) and (TH<=5):

			gamma=np.exp(5.75-1.15*TH)

	else:
			gamma=1

	if f<=fp:

    		sigma = 0.07

	else:

    		sigma = 0.09

	s=0.3125*Hs**2*Tp*(f/float(fp))**(-5)*np.exp(-1.25*(f/float(fp))**(-4))*(1-0.287*np.log(gamma))*gamma**(np.exp(-0.5*((f/float(fp)-1)/float(sigma))**2))

	return s

def x0(TDur,dt,Hs,Tp,f0,df,fHighCut,y,ny): # etaBC generation

	f=np.arange(f0,fHighCut+f0,df)

	omega=2*np.pi*f

	JS = np.zeros((1,f.size))

	for ij in range(0,f.size-1):

		fL=f[ij]

		JS1 = jonswap(fL,Hs,Tp)

		JS[0,ij] = JS1

	#tSpan=0:dt:TDur;
	tSpan=np.arange(0,TDur+dt,dt)

	etaBC=np.zeros((tSpan.size,1))

	A=np.zeros((1,omega.size))
	B=np.zeros((1,omega.size))
	AB=np.random.standard_normal((1,2*omega.size)) #random Fourie coefficients

	for j in range(0,omega.size):

	    A[0,j]=np.sqrt(JS[0,j]*df)*AB[0,j]
	    B[0,j]=np.sqrt(JS[0,j]*df)*AB[0,omega.size+j]

	for i in range(0,tSpan.size):

		etaBC[i,0]=np.sum(A*np.cos(2*np.pi*f*tSpan[i])+B*np.sin(2*np.pi*f*tSpan[i])); # Initial Wave Surface

	# Print Check
	#print omega.size

	#print etaBC
	#plt.plot(tSpan,etaBC,'b-*')
	#plt.xlabel('Time[s]')
	#plt.ylabel('Eta[m]')
	#plt.grid()
	#plt.show()

	n = tSpan.size

	# Lets print output the Wave Maker

	file = open("waveMakerSignalrandom.inp","w")

	file.write("# This is a Wave Generation surface response at x=0\n")
	file.write("%.7f %.0f %.0f\n" % (dt, n, ny))
	file.write("%.7f\n" % y)

	for i in range(0,tSpan.size):
	  file.write("%.7f %.7f\n" % (tSpan[i],etaBC[i,0]))

	file.close()

	## Which Fourie coefficients have been used

	file = open("randomAB.txt","w")

	file.write("# These are random Aj and Bj\n")

	for i in range(0,omega.size):
	  file.write("%.7s %.7s\n" % (AB[0,i],AB[0,omega.size+i]))

	file.close()

File: UQ-Packages/wavegen.m
function [etaBC] = wavegen(input,TDur,dt,Hs,Tp,f0,fHighCut,on)
%% Generate initial surface elevation based on H_S and T_P for JONSWAP
%{
---------------------------------------------------------------------------
Created by:
Kenan Šehić (kense@dtu.dk; kenosehic@gmail.com)
Department of Applied Mathematics and Computer Science
Technical University of Denmark
Licence: Copyright (C) 2019 Kenan Šehić DTU Compute, Technical University of Denmark

Cite: Šehić K., Bredmose H., Sørensen J.D., Karamehmedović M.: Low-dimensional representation of wave generation to quantify extreme events, TBD
Status: Submitted - Journal of Engineering Mathematics Dec 2019
---------------------------------------------------------------------------
Version December 2019
---------------------------------------------------------------------------
%}
%% Procedure
df=1/TDur; %frequ step

f=f0:df:fHighCut;

omega=2*pi*f;

%% Time
tSpan=0:dt:TDur;

%% Jonswap Spectrum

JS=zeros(1,length(f));

for ij=1:length(f)
   
    JS(1,ij)= jonswap(f(ij),Hs,Tp);
    
end

%% Wave generation

A=zeros(1,length(omega));
B=zeros(1,length(omega));
    
for j=1:length(omega)

    A(1,j)=sqrt(JS(1,j)*df).*input(j);
    B(1,j)=sqrt(JS(1,j)*df).*input(length(omega)+j);

end

etaBC(:,1)=sum(A(1,:).*cos(2*pi.*f.*tSpan')+B(1,:).*sin(2*pi.*f.*tSpan'),2); 

%% Making file

if on==1
    
    filename = 'initial_waves';

    Folder = pwd;

    FileName   = fullfile(Folder, sprintf('%s.iwf', filename));
    [fid, msg] = fopen(FileName, 'w');

    if fid == -1
        error('Cannot open file: %s', msg);
    end

    fprintf(fid, '# Time series generator\n');  % [EDITED]

    fprintf(fid, ['%g\n', ...
                  '%g\n'], dt,etaBC(1:end-1));  % [EDITED]
    fclose(fid);

end

return


